internal/datastore/mysql/datastore.go

```diff
@@ -0,0 +1,648 @@
+	createBaseTxn := fmt.Sprintf("INSERT IGNORE INTO %s (id) VALUES (1)", driver.RelationTupleTransaction())
+
+	gcCtx, cancelGc := context.WithCancel(context.Background())
+	querySplitter := common.TupleQuerySplitter{
+		Executor:         newMySQLExecutor(db),
+		UsersetBatchSize: config.splitAtUsersetCount,
+	}
+
+	store := &Datastore{
+		db:                       db,
+		driver:                   driver,
+		url:                      url,
+		revisionFuzzingTimedelta: config.revisionFuzzingTimedelta,
+		gcWindowInverted:         -1 * config.gcWindow,
+		gcInterval:               config.gcInterval,
+		gcMaxOperationTime:       config.gcMaxOperationTime,
+		gcCtx:                    gcCtx,
+		cancelGc:                 cancelGc,
+		watchBufferLength:        config.watchBufferLength,
+		createTxn:                createTxn,
+		createBaseTxn:            createBaseTxn,
+		QueryBuilder:             queryBuilder,
+		querySplitter:            &querySplitter,
+	}
+
+	// Start a goroutine for garbage collection.
+	if store.gcInterval > 0*time.Minute {
+		store.gcGroup, store.gcCtx = errgroup.WithContext(store.gcCtx)
+		store.gcGroup.Go(store.runGarbageCollector)
+	} else {
+		log.Warn().Msg("garbage collection disabled in mysql driver")
+	}
+
+	return store, nil
+}
+
+// unlike PSQL, this implementation does not create a transaction - not needed
+// prepared statements are also not used given they perform poorly
+// on environments where connections have short lifetime (e.g.
+// to gracefully handle load-balancer connection drain)
+func newMySQLExecutor(db *sql.DB) common.ExecuteQueryFunc {
+	return func(ctx context.Context, revision datastore.Revision, sqlQuery string, args []interface{}) ([]*core.RelationTuple, error) {
+		ctx = datastore.SeparateContextWithTracing(ctx)
+
+		span := trace.SpanFromContext(ctx)
+
+		rows, err := db.QueryContext(ctx, sqlQuery, args...)
+		if err != nil {
+			return nil, fmt.Errorf(errUnableToQueryTuples, err)
+		}
+		defer migrations.LogOnError(ctx, rows.Close)
+
+		span.AddEvent("Query issued to database")
+
+		var tuples []*core.RelationTuple
+		for rows.Next() {
+			nextTuple := &core.RelationTuple{
+				ObjectAndRelation: &core.ObjectAndRelation{},
+				User: &core.User{
+					UserOneof: &core.User_Userset{
+						Userset: &core.ObjectAndRelation{},
+					},
+				},
+			}
+			userset := nextTuple.User.GetUserset()
+			err := rows.Scan(
+				&nextTuple.ObjectAndRelation.Namespace,
+				&nextTuple.ObjectAndRelation.ObjectId,
+				&nextTuple.ObjectAndRelation.Relation,
+				&userset.Namespace,
+				&userset.ObjectId,
+				&userset.Relation,
+			)
+			if err != nil {
+				return nil, fmt.Errorf(errUnableToQueryTuples, err)
+			}
+
+			tuples = append(tuples, nextTuple)
+		}
+		if err := rows.Err(); err != nil {
+			return nil, fmt.Errorf(errUnableToQueryTuples, err)
+		}
+		span.AddEvent("Tuples loaded", trace.WithAttributes(attribute.Int("tupleCount", len(tuples))))
+		return tuples, nil
+	}
+}
+
+type Datastore struct {
```

Comment:
All exported members must have doc strings

Reply:
Happy to add them! If this is a requirement in the `internal` package, shouldn't this enabled via `golangci-lint`?

Reply:
addressed in https://github.com/authzed/spicedb/pull/525/commits/7161a836868da484c9c5824d14ca61681ffdb5a3

Reply:
apparently `golint`, the linter that enforced this, is now deprecated. `golangci-lint` offers an alternative but I'm not sure if that one enforces exported type comments.

Addressed in https://github.com/authzed/spicedb/pull/525/commits/027ddec640ba1fd147938b39dce08bfa2894c7aa